lib/ssl/src/tls_connection_1_3.erl

```diff
@@ -237,61 +237,94 @@ user_hello({call, From}, cancel, State) ->
     ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?USER_CANCELED, user_canceled),
                                     ?FUNCTION_NAME, State);
 user_hello({call, From}, {handshake_continue, NewOptions, Timeout},
-           #state{static_env = #static_env{role = Role},
-                  handshake_env = #handshake_env{hello = Hello},
+           #state{static_env = #static_env{role = client = Role},
+                  handshake_env = HSEnv,
                   ssl_options = Options0} = State0) ->
-    Options = ssl:handle_options(NewOptions, Role, Options0#{handshake => full}),
+    Options = ssl:handle_options(NewOptions, Role, Options0),
     State = ssl_gen_statem:ssl_config(Options, Role, State0),
-    Next = case Role of
-               client ->
-                   wait_sh;
-               server ->
-                   start
-           end,
-    {next_state, Next, State#state{start_or_recv_from = From}, 
-     [{next_event, internal, Hello}, {{timeout, handshake}, Timeout, close}]};
+    {next_state, wait_sh, State#state{start_or_recv_from = From,
+                                      handshake_env = HSEnv#handshake_env{handshake_continue = continue}},
+     [{{timeout, handshake}, Timeout, close}]};
+user_hello({call, From}, {handshake_continue, NewOptions, Timeout},
+           #state{static_env = #static_env{role = server = Role},
+                  handshake_env = #handshake_env{handshake_continue = ClientVersions} = HSEnv,
+                  ssl_options = Options0} = State0) ->
+    Options = #{versions := Versions} = ssl:handle_options(NewOptions, Role, Options0),
+    State = ssl_gen_statem:ssl_config(Options, Role, State0),
+    case ssl_handshake:select_supported_version(ClientVersions, Versions) of
+        {3,4} ->
+            {next_state, start, State#state{start_or_recv_from = From,
+                                            handshake_env = HSEnv#handshake_env{handshake_continue = continue}},
+             [{{timeout, handshake}, Timeout, close}]};
+        undefined ->
+            ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?PROTOCOL_VERSION), ?FUNCTION_NAME, State);
+        _Else ->
+            {next_state, hello, State#state{start_or_recv_from = From,
+                                            handshake_env = HSEnv#handshake_env{handshake_continue = continue}},
+             [{change_callback_module, tls_connection},
+              {{timeout, handshake}, Timeout, close}]}
+    end;
```

Comment:
When does this happen?

Reply:
This is a corner case that could happen for the server if the user changes the version option and client happens to work a certain way.